keps/sig-auth/740-service-account-external-signing/README.md

```diff
@@ -1,113 +1,605 @@
-  bytes payload = 1;
-  // algorithm specifies which algorithm to sign with
-  string algorithm = 2;
+message SignJWTResponse {
+  // header must contain alg, kid, typ claims.
+  // type must be “JWT”.
+  // kid must be non-empty.
+  // alg must be one of the algorithms supported by kube-apiserver (currently RS256, ES256, ES384, ES512).
+  // header cannot have any additional data that kube-apiserver does not recognize.
+  // Already wrapped in URL-safe base64, exactly as it appears in the first segment of the JWT.
+  string header = 1;
+
+  // The signature for the JWT.
+  // Already wrapped in URL-safe base64, exactly as it appears in the final segment of the JWT.
+  string signature = 2;
 }
-message SignPayloadResponse {
-  // content returns the signed payload
-  bytes content = 1;
+
+message FetchKeysRequest {}
+
+message FetchKeysResponse {
+  repeated Key keys = 1;
+
+  // The timestamp when this data was pulled from the authoritative source of
+  // truth for verification keys.
+  // kube-apiserver can export this from metrics, to enable end-to-end SLOs.
+  google.protobuf.Timestamp data_timestamp = 2;
+
+  // refresh interval for verification keys to pick changes if any.
+  // any value <= 0 is considered a misconfiguration.
+  int64 refresh_hint_seconds = 3;
 }
 
+message Key {
+  // A unique identifier for this key.
+  string key_id = 1;
+
+  // The public key, PKIX-serialized.
+  bytes key = 2;
 
-message PublicKey {
-  // public_key is a PEM encoded public key
-  bytes public_key = 1;
-  // certificate is a concatenated list of PEM encoded x509 certificates
-  bytes certificates = 2;
-  // key_id is the key's ID
-  string key_id = 3;
-  // algorithm states the algorithm the key uses
-  string algorithm = 4;
+  // Set only for keys that are not used to sign bound tokens.
+  // eg: supported keys for legacy tokens.
+  // If set, key is used for verification but excluded from OIDC discovery docs.
+  bool exclude_from_oidc_discovery = 3;
 }
 
-message ListPublicKeysRequest {}
-message ListPublicKeysResponse {
-  // key_id is the key's ID
-  string active_key_id = 1;
-  // public_keys is a list of public verifying keys
-  repeated PublicKey public_keys = 2;
+message MetadataRequest {}
+
+message MetadataResponse {
+  // used by kube-apiserver for defaulting/validation of JWT lifetime while accounting for configuration flag values:
+  // 1. `--service-account-max-token-expiration`
+  // 2. `--service-account-extend-token-expiration`
+  //
+  // * If `--service-account-max-token-expiration` is greater than `max_token_expiration_seconds`, kube-apiserver treats that as misconfiguration and exits.
+  // * If `--service-account-max-token-expiration` is not explicitly set, kube-apiserver defaults to `max_token_expiration_seconds`.
+  // * If `--service-account-extend-token-expiration` is true, the extended expiration is `min(1 year, max_token_expiration_seconds)`.
```

Comment:
This seems like a weird interaction that will surprise people when it happens.

Reply:
Are you suggesting that it should rather result in an validation error?

cc: @liggitt

Reply:
IIRC, @deads2k wanted the external signer to be able to control the ultimate expirations without the kube-apiserver args needing to change. The alternatives are:
1. fail to start the kube-apiserver
2. silently default `--service-account-extend-token-expiration` to false
3. try to get the external signer to sign year-long extended duration tokens anyway
4. make the extended token duration as long as possible while fitting in the bounds of the external signer

I think 4 is the least surprising

Reply:
My comment was mostly around 'how do we make it clear to folks that tokens that used to be extended to a year now are extended to only a week' _before_ they have workloads that depend on the extension start breaking seemingly at random (i.e. the pod has been running long enough).

I agree that 4 seems like an okay option given the constraints.

Reply:
I'd put that in the category of things the admin / integrator should figure out before switching to an external token signer that will force token lifetimes to be shorter. Good to make a note of to document the specific metrics they can check to determine if extended token lifetimes are being relied on in the cluster (the stale tokens metric)

Reply:
called it out in a new section under "monitoring requirements"  ("Other useful metrics that can assist with this feature")